### ISSUE TITLE
[Typscript] Wrong type for `new Model(doc)`

### ISSUE THREAD

**Issue by Zenthae** [Automattic/mongoose#10302]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**

When using `new Model(doc)` to create a new model and the `save()` it. the type of `doc` is always `any` which break the auto-completion for properties added by the user model.

![image](https://user-images.githubusercontent.com/32647899/119918891-9065e400-bf69-11eb-918d-cab2e65e34b5.png)

**If the current behavior is a bug, please provide the steps to reproduce.**

[Quick Start](https://mongoosejs.com/docs/index.html) example from the website

`tsconfig.json`
```json
{
  "compilerOptions": {
    "target": "ES2020",
    "module": "commonjs",
    "outDir": "./dist",
    "rootDir": "./src",
    "strict": true,
    "moduleResolution": "node",
    "types": ["mongoose"],
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true
  },
  "exclude": ["node_modules"],
  "include": ["src/**/*.ts"]
}
```

**What is the expected behavior?**

instead of having `any` it should use the provided type.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

nodejs: v16.1.0
mongoose: 5.12.10

**Comment by DVGY** [comment]
const Kitten = mongoose.model<IKitten>('Kitten', kittenSchema);

Is this line added correctly in your code ?

**Comment by Zenthae** [comment]
yes it is, it's the line hidden by Vscode popup

**Comment by DVGY** [comment]
> yes it is, it's the line hidden by Vscode popup
@Zenthae change the line 
from this

`const Kitten = mongoose.model('Kitten', kittenSchema);`

to this
`const Kitten = mongoose.model<Ikitten>('Kitten', kittenSchema);`

**Comment by Zenthae** [comment]
same

**Comment by DVGY** [comment]
> same

@Zenthae one more thing make sure your `interface IKitten extends Document` is like this

**Comment by Zenthae** [comment]
still the same issue

**Comment by DVGY** [comment]
> still the same issue

can you share on codepen ?

**Comment by vkarpov15** [comment]
Make sure you're running Mongoose >= 5.12.6, we changed this type definition with #10074 so either you're using an older version of Mongoose or VSCode is picking up the wrong version of Mongoose type definitions.

**Comment by Zenthae** [comment]
i'm using mongoose `5.12.13`, checked the type definition, 
```ts
.....
interface Model<T, TQueryHelpers = {}, TMethods = {}> extends NodeJS.EventEmitter, AcceptsDiscriminator {
    new(doc?: T | any): EnforceDocument<T, TMethods>;
.....
```
but it's still not working, the inferred type is `any` for `new()`

here is my model
```ts
interface UserModel extends IUser, Document {
  comparePassword(plainPassword: string): boolean;
}

export const userSchema = new Schema<UserModel>({
  username: { type: String, unique: true },
  password: String,
});

userSchema.pre('save', function (this: UserModel, next) {
  if (!this.isModified('password')) return next();

  this.password = hashSync(this.password, 15);

  next();
});

userSchema.method('comparePassword', function (this: UserModel, plainPassword: string) {
  return compareSync(plainPassword, this.password);
});

export const User = model<UserModel>('User', userSchema);

const u = new User({});
```
i tested using [this](https://github.com/Automattic/mongoose/pull/10165/files#diff-f84f70bda667ab995975c042afd7fc327837b886b5e0088eaf17748e5f9a036eR68-R90) too, and it's still inferring `any`, so maybe it's my vscode that doing something wrong but then i don't now how to even start fixing or testing it ....

**Comment by vkarpov15** [comment]
We should see if:

```typescript
new(doc?: T): EnforceDocument<T, TMethods>;
new(doc?: any): EnforceDocument<T, TMethods>;
```

helps, re: https://github.com/Automattic/mongoose/pull/10343

**Comment by vkarpov15** [comment]
We took a look and unfortunately the suggestion in https://github.com/Automattic/mongoose/issues/10302#issuecomment-865185892 doesn't really help. TypeScript just always falls back to `any` if any required properties in `T` are missing. You should use `as User` if you want to get better VS code autocomplete:

```typescript
interface User {
  name: string;
  id: number;
}

const UserModel = model<User>('User', new Schema({ name: String, id: Number }));

const doc = new UserModel({ name: 'test' } as User);
```

An alternative that we'll consider for the future is making `UserModel()` strongly typed for better autocomplete, but leaving `new UserModel()` the way it currently is. So the below TypeScript defs:

```typescript
    new(doc?: T | any): EnforceDocument<T, TMethods>;
    (doc?: T): EnforceDocument<T, TMethods>;
```

Another potential alternative would be creating a separate static function, like `from()`, that allows for better autocomplete. Like `UserModel.from({ ... })`

**Comment by Zenthae** [comment]
i think that the `from()` solution hold value, because at the only way, (from my knowledge, which aren't lot) to create a new Model instance is the `create()` function which automatically add it to the database, which might not be the desired outcome

**Comment by cakeslice** [comment]
I think there should be a way to enforce object creation that matches the schema.

``const doc = new UserModel({ name: 'test' } as User);`` is error prone.
All it takes is to forget the "as User" and then:
`const doc = new UserModel({ namee: 'test' });` goes through.

Not to mention having to add the "as User" everywhere is not practical.

**Comment by FINDarkside** [comment]
> An alternative that we'll consider for the future is making UserModel() strongly typed for better autocomplete, but leaving new UserModel() the way it currently is.

Is there some reason why anyone would want it to not be strongly typed? This is kinda a bummer since AFAIK we had strongly typed constructor params and create function until mongoose started providing its own types and `@types/mongoose` wasn't maintained anymore.

**Comment by vkarpov15** [comment]
@FINDarkside "Is there a reason why someone would not want it to be strongly typed?" <-- Yes, if you want to pass in `req.body` or some other arbitrary object.

**Comment by FINDarkside** [comment]
It'd still work if `req.body` is `any` and not `unknown`. If it's unknown you can just cast it to `any`. It'd be breaking change for sure, but the fix would be trivial. But didn't really consider that `unknown` technically would be valid type for it since mongoose does the validation, but personally I don't feel like disallowing `unknown` would be a big problem.

**Comment by mjfwebb** [comment]
> @FINDarkside "Is there a reason why someone would not want it to be strongly typed?" <-- Yes, if you want to pass in `req.body` or some other arbitrary object.

I'm not sure about your reasoning here @vkarpov15. This is a confusing reason to default to `any`.

From a user of Typescript perspective, I'd want `new MyModel` to expect input parameters as defined in the `interface` I had explicitly associated with the model.

While there may well be users that want to pass in req.body or an arbitrary object, I'd argue that they should be the ones to explicitly code in that intention.

**Comment by andrewrjohn** [comment]
> From a user of Typescript perspective, I'd want new MyModel to expect input parameters as defined in the interface I had explicitly associated with the model.

This is exactly what I'd expect to be getting from Mongoose when using Typescript (and I think most people would agree). Does anybody know if there is progress being made on this?

**Comment by hasezoey** [comment]
is this still a issue with mongoose 6.6.1?

currently `new` types are:
https://github.com/Automattic/mongoose/blob/69c99d2d9ee4d13f1635eaa63d042f3a6a6fb7a3/types/models.d.ts#L124

**Comment by vkarpov15** [comment]
@hasezoey oddly enough, I think this behavior is still there. I don't think this behavior is an issue though.

For example, the following script compiles fine.

```ts
import mongoose from 'mongoose';

const schema = new mongoose.Schema<{ name?: string }>({ name: String });
const Test = mongoose.model<{ name?: string }>('Test', schema);

const doc = new Test({ name: 42, other: 'bar' });
```

And VS Code autocomplete thinks that `DocType` is `{}` ?

![image](https://user-images.githubusercontent.com/1620265/193463136-36fb1a62-5ecf-43a9-815b-5700b4380291.png)

I genuinely am not sure why.

**Comment by hasezoey** [comment]
> And VS Code autocomplete thinks that DocType is {} ?
> I genuinely am not sure why.

well in this case it is because the parameter `doc` uses the function generic `DocType`, which is defaulted to a unbound `T` generic in `Model`, and because it is a "creating method" (how i call it), it uses the value it finds as that generic

TL;DR: the type for parameter `doc` is unbound, so it uses the value it has, which is `{}`

generic `T`:
https://github.com/Automattic/mongoose/blob/ff7eed50ead64d209b9d33a6b4fff58d41d7c3b1/types/models.d.ts#L119

generic `DocType` (and the whole function declaration):
https://github.com/Automattic/mongoose/blob/ff7eed50ead64d209b9d33a6b4fff58d41d7c3b1/types/models.d.ts#L124

**Comment by FilipPyrek** [comment]
Hi there,

I just entered mongoose world and I want it to use in my TypeScript project . So far the setup was okay... but I found out same issue as @hasezoey is talking about in his latest message.

Wrong mongoose type definition in `Model`:
```typescript
new <DocType = T>(doc?: DocType, fields?: any | null, options?: boolean | AnyObject): HydratedDocument<T, TMethodsAndOverrides, IfEquals<TVirtuals, {}, ObtainSchemaGeneric<TSchema, 'TVirtuals'>, TVirtuals>> & ObtainSchemaGeneric<TSchema, 'TStaticMethods'>;
```

produces incorrect behavior when creating new instance of a model
![image](https://user-images.githubusercontent.com/6282843/195433748-a483ceaf-03c5-4993-a23f-06e90f33e08a.png)
![image](https://user-images.githubusercontent.com/6282843/195433849-64caa84d-e715-4fba-901b-2d03fca11aec.png)
No type error message coming from TypeScript even though the input doesn't comply with the schema ☝️ 

When I fix it to 
```typescript
 new (doc?: T, fields?: any | null, options?: boolean | AnyObject): HydratedDocument<T, TMethodsAndOverrides, IfEquals<TVirtuals, {}, ObtainSchemaGeneric<TSchema, 'TVirtuals'>, TVirtuals>> & ObtainSchemaGeneric<TSchema, 'TStaticMethods'>;
```

Then I'm getting the correct TypeScript error:
![image](https://user-images.githubusercontent.com/6282843/195434307-83680e10-11d1-406a-aa12-e2a159988324.png)

If needed, I can contribute to help to solve this issue ASAP. 🙂

**Comment by hasezoey** [comment]
> If needed, I can contribute to help to solve this issue ASAP. slightly_smiling_face

the change you have made would conflict with other things mongoose supports, at least from what i can tell: because before a `save` (/ `validate`) values can be added / changed and are not required to be set in `new Model` directly, but this is a different story for `Model.create`.

the proper thing for this issue is to bind `T` to something instead of being unbound

**Comment by FilipPyrek** [comment]
Well, that's quite unfortunate because it breaks Developer Experience when trying to ensure type safety with TypeScript.

Maybe it could be done via conditional generic parameters like this for example:
```typescript
// just a quick illustration of the idea

 new <DocType = T>(doc?: DocType extends 'full' ? T : DocType, fields?: any | null, options?: boolean | AnyObject): HydratedDocument<T, TMethodsAndOverrides, IfEquals<TVirtuals, {}, ObtainSchemaGeneric<TSchema, 'TVirtuals'>, TVirtuals>> & ObtainSchemaGeneric<TSchema, 'TStaticMethods'>;
```
<img width="339" alt="image" src="https://user-images.githubusercontent.com/6282843/195584561-68abb558-6450-4f0d-9e34-8fec961e1548.png">

**Comment by hasezoey** [comment]
i dont think it is a good way to use `DocType` itself, but ok at showcasing what you mean
@vkarpov15 @AbdelrahmanHafez what do you think about this proposed change (conditional generic for strict types)?

**Comment by vkarpov15** [comment]
@FilipPyrek how exactly is your suggested approach different from what is already there? The `DocType` generic is already `T` by default, so if you don't set any generic you should get the same result.

**Comment by FilipPyrek** [comment]
> @FilipPyrek how exactly is your suggested approach different from what is already there? The `DocType` generic is already `T` by default, so if you don't set any generic you should get the same result.

That suggested solution doesn't create a schema breaking change. The problem is that `DocType` is actually never using the default type `T`, because `DocType` type is always available (even if it's just of type `{}`), thus you never get to the default `T` type.

In the suggested solution you can choose if you want the original "broken" approach or if you want `doc` to be always of type `T`, thus get correct TypeScript checks.

**Comment by JavaScriptBach** [comment]
https://github.com/Automattic/mongoose/pull/13038 fixes this for `Model.create()`. @vkarpov15 @hasezoey would you be willing to accept this patch?

**Comment by akraminakib** [comment]
This is still an issue with v8 of mongoose even when following the examples from [their docs](https://mongoosejs.com/docs/typescript/statics-and-methods.html). Running the below lines infers the results on each to be of type `any`:

```ts
const user1 = await User.findOne({}).exec();

// where params is an object that represents the valid fields defined on the user schema
const user2 = new User({...params});
```
